fluentd: add write timeout log option

aaithal opened the pull request:
Added support for `fluentd-write-timeout` configuration option in the fluentd logger to prevent indefinite blocking when downstream connections become unhealthy. This allows fluentd to abandon unresponsive connections after a specified duration instead of being blocked forever.

Key changes:
- Added new `writeTimeoutKey` configuration option
- Implemented duration parsing and validation
- Added unit tests to verify timeout behavior with unresponsive server
- Default timeout remains 0 (no timeout) for backward compatibility

**- What I did**
Modified `fluentd` logdriver code to expose a new log option. The log option can be used to specify write timeouts.  

**- How I did it**
The code change exposes the existing config to specify a write timeout value for writes to fluentd server.

**- How to verify it**
Added a unit test, which fails when the timeout is not specified. The test succeeds when a timeout value is specified.

**- Human readable description for the release notes**
```markdown changelog
Add a new log option for fluentd log driver (`fluentd-write-timeout`), which enables specifying write timeouts for fluentd connections.
```

aaithal committed:
fluentd: add write timeout log option

Currently, there's no mechanism to specify a write timeout value for
fluentd connections. This means that writes can forever be blocked if
the downstream connections is unhealthy. This commit makes this value
configurable via a new fluentd log option called "fluentd-write-timeout".

Signed-off-by: Anirudh Aithal <aithal@amazon.com>

austinvazquez reviewed (approved):
+1, thanks for adding the tests.

thaJeztah commented on daemon/logger/fluentd/fluentd_test.go:
```diff
@@ -199,3 +205,118 @@ func TestValidateLogOptAddress(t *testing.T) {
+		}
+	}
+
+	assert.Check(t, err != nil)
```
Perhaps good to check if we got the expected error here; 


```go
assert.Error(t, err, "fluent#write: failed to write after 1 attempts")
```

aaithal replied:
The challenge is that depending on the configuration of the underlying machine where the test is being run, it may some times fail in the 2nd or the 3rd attempt. I could may be do a pattern matching on the error message to ensure that the error that's expected is returned.

thaJeztah commented on daemon/logger/fluentd/fluentd_test.go:
```diff
@@ -199,3 +205,118 @@ func TestValidateLogOptAddress(t *testing.T) {
+	assert.NilError(t, err, "unable to create listener for socket %s", socketFile)
+	defer l.Close()
+
+	ctx, cancel := context.WithCancel(context.Background())
```
I was trying the test with the option disabled, and that caused it (expectedly) to hang forever; perhaps we should add a timeout here .. just in case?


```go
ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```

aaithal replied:
sure, sounds good.

thaJeztah commented on daemon/logger/fluentd/fluentd_test.go:
```diff
@@ -199,3 +205,118 @@ func TestValidateLogOptAddress(t *testing.T) {
+
+			// Handle an incoming connection. We're essentially blackholing this by not reading from
+			// or writing to the connection.
+			go func(ctx context.Context, conn net.Conn) {
+				defer conn.Close()
+				// Simulate unresponsive server: do nothing with the connection
+				<-ctx.Done()
+			}(ctx, conn)
```
Micro-nit; can remove the defer, to make my IDE ever-so-slightly less noisy about "unhandled error" 😂 

```go
go func(ctx context.Context, conn net.Conn) {
	// Simulate unresponsive server: do nothing with the connection
	<-ctx.Done()
	_ = conn.Close()
}(ctx, conn)
```

thaJeztah commented on daemon/logger/fluentd/fluentd.go:
```diff
@@ -237,6 +243,18 @@ func parseConfig(cfg map[string]string) (fluent.Config, error) {
+	writeTimeout := time.Duration(0)
+	if cfg[writeTimeoutKey] != "" {
+		if d, err := time.ParseDuration(cfg[writeTimeoutKey]); err != nil {
+			return config, errors.Wrapf(err, "invalid value for %s: value must be a duration", writeTimeoutKey)
+
```
Stray empty line here

thaJeztah commented on daemon/logger/fluentd/fluentd_test.go:
```diff
@@ -199,3 +205,118 @@ func TestValidateLogOptAddress(t *testing.T) {
+		t.Run("valid "+k, func(t *testing.T) {
+			err := ValidateLogOpt(map[string]string{writeTimeoutKey: k})
+			assert.NilError(t, err)
+			// Ignore err here since that's already checked in the ValidateLogOpt error check.
+			cfg, _ := parseConfig(map[string]string{writeTimeoutKey: k})
```
1-line comment to explain why we're not adding a 1-line assert; I guess we might as well just assert it's indeed not an error;

```go
cfg, err := parseConfig(map[string]string{writeTimeoutKey: k})
assert.NilError(t, err)
```

thaJeztah reviewed (changes requested):
thanks! overall looking good; I left some nits and suggestions, but let me know what you think!

aaithal commented:
Thanks for reviewing this and for the comments. I can address a bulk of these and post an update.

thaJeztah commented:
Thanks! Yes, give me a ping when you had time to do so; when doing so, please amend the commit (squashed); no need to keep them as separate commits ❤️

thaJeztah commented on daemon/logger/fluentd/fluentd_test.go:
```diff
@@ -199,3 +206,127 @@ func TestValidateLogOptAddress(t *testing.T) {
+		t.Run("valid "+k, func(t *testing.T) {
+			err := ValidateLogOpt(map[string]string{writeTimeoutKey: k})
+			assert.NilError(t, err)
+			// Ignore err here since that's already checked in the ValidateLogOpt error check.
```
oh! forgot to remove the comment that's now no longer the case 🙈 😅

thaJeztah commented on daemon/logger/fluentd/fluentd_test.go:
```diff
@@ -199,3 +206,127 @@ func TestValidateLogOptAddress(t *testing.T) {
+	assert.Check(t, err != nil)
+	// Check if error string matches the pattern "fluent#write: failed to write after %d attempts"
+	// where %d is a number.
+	assert.Check(t, cmp.Regexp(`fluent#write: failed to write after \d+ attempts`, err.Error()))
```
It's fine to simplify to just "error contains"; checking that prefix should be "good enough" to know we have the expected error;

```go
assert.ErrorContains(t, err, "fluent#write: failed to write after")
```

When doing so, you can also remove the `assert.Check(t, err != nil)` above (which is then redundant)

aaithal replied:
Yeah, seems like `ErrorContains` does a `err != nil` check as well -> https://github.com/gotestyourself/gotest.tools/blob/0b81523ff268a1f1b0baf4a5da67e42fbb86880b/assert/cmp/compare.go#L233

Will modify this.

aaithal commented:
I think I've addressed all of the comments with this change: https://github.com/moby/moby/compare/db52f6b42724419a5f2140c740d5c87eae8f4534..7dae7c54dde81951330170c7678deea350b1d2b8. Can you please take another look @thaJeztah ? Thanks in advance!

thaJeztah reviewed (approved):
LGTM, thanks!
